File: WFM-Planning-Suite/src/core/datetime/parser.py
"""Single source of truth for date parsing across the pipeline."""

from __future__ import annotations

import re
from datetime import datetime
from enum import Enum


class DateOrder(str, Enum):
    DMY = "DMY"
    MDY = "MDY"
    YMD = "YMD"


_FORMATS: dict[DateOrder, list[str]] = {
    DateOrder.DMY: ["%d/%m/%Y %H:%M", "%d-%m-%Y %H:%M", "%d/%m/%Y", "%d-%m-%Y"],
    DateOrder.MDY: ["%m/%d/%Y %H:%M", "%m-%d-%Y %H:%M", "%m/%d/%Y", "%m-%d-%Y"],
    DateOrder.YMD: ["%Y-%m-%d %H:%M", "%Y/%m/%d %H:%M", "%Y-%m-%d", "%Y/%m/%d"],
}


def _has_unambiguous_day(value: str) -> bool:
    for sep in ("/", "-"):
        parts = value.split(sep)
        if len(parts) >= 2:
            try:
                first = int(parts[0])
                second = int(parts[1])
                if first > 12 or second > 12:
                    return True
            except ValueError:
                continue
    return False


def _infer_order_from_unambiguous(values: list[str]) -> DateOrder | None:
    dmy_score = 0
    mdy_score = 0
    ymd_score = 0

    for value in values:
        if not value or not value.strip():
            continue
        text = value.strip()
        if re.match(r"^\d{4}", text):
            ymd_score += 2
            continue
        if not _has_unambiguous_day(text):
            continue
        for sep in ("/", "-"):
            parts = text.split(sep)
            if len(parts) < 2:
                continue
            try:
                first = int(parts[0])
                second = int(parts[1])
            except ValueError:
                continue
            if first > 12:
                dmy_score += 1
            if second > 12:
                mdy_score += 1

    if ymd_score > max(dmy_score, mdy_score):
        return DateOrder.YMD
    if dmy_score > mdy_score:
        return DateOrder.DMY
    if mdy_score > dmy_score:
        return DateOrder.MDY
    return None


def resolve_date_format(values: list[str], default: DateOrder = DateOrder.DMY) -> tuple[DateOrder, str]:
    """Resolve date convention once from the whole file using unambiguous rows."""
    order = _infer_order_from_unambiguous(values)
    if order is None:
        order = default
    fmt = _FORMATS[order][0]
    return order, fmt


def parse_datetime(value: str, order: DateOrder) -> datetime:
    text = value.strip()
    for fmt in _FORMATS[order]:
        try:
            return datetime.strptime(text, fmt)
        except ValueError:
            continue
    raise ValueError(f"Cannot parse date '{value}' with order {order.value}")


def parse_series(values: list[str], order: DateOrder | None = None) -> tuple[list[datetime], DateOrder, str]:
    non_empty = [v for v in values if v and str(v).strip()]
    if not non_empty:
        raise ValueError("No date values to parse")
    resolved_order, fmt = resolve_date_format(non_empty) if order is None else (order, _FORMATS[order][0])
    parsed = [parse_datetime(v, resolved_order) for v in values if v and str(v).strip()]
    return parsed, resolved_order, fmt

File: WFM-Planning-Suite/web/src/core/datetime/parser.ts
export type DateOrder = 'DMY' | 'MDY' | 'YMD'

const FORMATS: Record<DateOrder, string[]> = {
  DMY: ['DD/MM/YYYY HH:mm', 'DD-MM-YYYY HH:mm', 'DD/MM/YYYY', 'DD-MM-YYYY'],
  MDY: ['MM/DD/YYYY HH:mm', 'MM-DD-YYYY HH:mm', 'MM/DD/YYYY', 'MM-DD-YYYY'],
  YMD: ['YYYY-MM-DD HH:mm', 'YYYY/MM/DD HH:mm', 'YYYY-MM-DD', 'YYYY/MM/DD'],
}

function hasUnambiguousDay(value: string): boolean {
  for (const sep of ['/', '-']) {
    const parts = value.split(sep)
    if (parts.length >= 2) {
      const first = parseInt(parts[0], 10)
      const second = parseInt(parts[1], 10)
      if (!isNaN(first) && !isNaN(second) && (first > 12 || second > 12)) return true
    }
  }
  return false
}

export function resolveDateFormat(values: string[], defaultOrder: DateOrder = 'DMY'): [DateOrder, string] {
  let dmy = 0, mdy = 0, ymd = 0
  for (const value of values) {
    const text = value.trim()
    if (!text) continue
    if (/^\d{4}/.test(text)) { ymd += 2; continue }
    if (!hasUnambiguousDay(text)) continue
    for (const sep of ['/', '-']) {
      const parts = text.split(sep)
      if (parts.length < 2) continue
      const first = parseInt(parts[0], 10)
      const second = parseInt(parts[1], 10)
      if (isNaN(first) || isNaN(second)) continue
      if (first > 12) dmy++
      if (second > 12) mdy++
    }
  }
  let order: DateOrder = defaultOrder
  if (ymd > Math.max(dmy, mdy)) order = 'YMD'
  else if (dmy > mdy) order = 'DMY'
  else if (mdy > dmy) order = 'MDY'
  return [order, FORMATS[order][0]]
}

function parseWithOrder(text: string, order: DateOrder): Date {
  const trimmed = text.trim()
  const patterns: Array<{ re: RegExp; fn: (m: RegExpMatchArray) => Date }> = []

  if (order === 'DMY') {
    patterns.push({
      re: /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})\s+(\d{1,2}):(\d{2})$/,
      fn: (m) => new Date(+m[3], +m[2] - 1, +m[1], +m[4], +m[5]),
    })
    patterns.push({
      re: /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})$/,
      fn: (m) => new Date(+m[3], +m[2] - 1, +m[1]),
    })
  } else if (order === 'MDY') {
    patterns.push({
      re: /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})\s+(\d{1,2}):(\d{2})$/,
      fn: (m) => new Date(+m[3], +m[1] - 1, +m[2], +m[4], +m[5]),
    })
    patterns.push({
      re: /^(\d{1,2})[\/\-](\d{1,2})[\/\-](\d{4})$/,
      fn: (m) => new Date(+m[3], +m[1] - 1, +m[2]),
    })
  } else {
    patterns.push({
      re: /^(\d{4})[\/\-](\d{1,2})[\/\-](\d{1,2})\s+(\d{1,2}):(\d{2})$/,
      fn: (m) => new Date(+m[1], +m[2] - 1, +m[3], +m[4], +m[5]),
    })
    patterns.push({
      re: /^(\d{4})[\/\-](\d{1,2})[\/\-](\d{1,2})$/,
      fn: (m) => new Date(+m[1], +m[2] - 1, +m[3]),
    })
  }

  for (const { re, fn } of patterns) {
    const m = trimmed.match(re)
    if (m) return fn(m)
  }
  throw new Error(`Cannot parse date '${text}' with order ${order}`)
}

export function parseSeries(values: string[]): [Date[], DateOrder, string] {
  const nonEmpty = values.filter((v) => v && v.trim())
  if (!nonEmpty.length) throw new Error('No date values to parse')
  const [order, fmt] = resolveDateFormat(nonEmpty)
  const parsed = nonEmpty.map((v) => parseWithOrder(v, order))
  return [parsed, order, fmt]
}
